src/lib/transition_router/transition_router.ml

```diff
@@ -139,6 +122,144 @@ module Make (Inputs : Inputs_intf) = struct
           ~transition_reader_ref ~transition_writer_ref ~frontier_w
           new_frontier )
 
+  let download_best_tip ~logger ~network ~verifier ~trust_system =
+    let num_peers = 8 in
+    let peers = Network.random_peers network num_peers in
+    Logger.info logger ~module_:__MODULE__ ~location:__LOC__
+      "Requesting peers for their best tip to do initialization" ;
+    Deferred.List.fold peers ~init:None ~f:(fun acc peer ->
+        match%bind Network.get_best_tip network peer () with
+        | Error e ->
+            Logger.debug logger ~module_:__MODULE__ ~location:__LOC__
+              ~metadata:[("peer", Network_peer.Peer.to_yojson peer)]
+              !"Could not get best tip from peer: %{sexp:Error.t}"
+              e ;
+            return acc
+        | Ok peer_best_tip -> (
+            match%bind Best_tip_prover.verify ~verifier peer_best_tip with
+            | Error e ->
+                let error_msg =
+                  sprintf
+                    !"Peer %{sexp:Network_peer.Peer.t} sent us bad proof for \
+                      their best tip"
+                    peer
+                in
+                Logger.warn logger ~module_:__MODULE__ ~location:__LOC__
+                  ~metadata:[("error", `String (Error.to_string_hum e))]
+                  !"%s" error_msg ;
+                let%map () =
+                  Trust_system.(
+                    record trust_system logger peer.host
+                      Actions.(Violated_protocol, Some (error_msg, [])))
+                in
+                acc
+            | Ok (`Root _, `Best_tip candidate_best_tip) ->
+                let enveloped_candidate_best_tip =
+                  Envelope.Incoming.wrap ~data:candidate_best_tip
+                    ~sender:(Envelope.Sender.Remote peer.host)
+                in
+                return
+                @@ Option.merge acc
+                     (Option.return enveloped_candidate_best_tip)
+                     ~f:(fun existing_best_tip candidate_best_tip ->
+                       if
+                         Consensus.Hooks.select
```

Comment:
For this comparison, you can just use `External_transition.max`. This compares transitions in terms of `Consensus.Hooks.select`.